iib/web/models.py

```diff
@@ -190,6 +204,79 @@ def add_state(self, state, state_reason):
 
+    def to_json(self, verbose=True):
+        states = [
+            {
+                'state': RequestStateMapping(state.state).name,
+                'state_reason': state.state_reason,
+                'updated': state.updated.isoformat(),
+            }
+            for state in self.states
+        ]
+        # Reverse the list since the latest states should be first
+        states = list(reversed(states))
+        latest_state = states[0]
+
+        rv = {
+            'id': self.id,
+            'bundles': [bundle.to_json() for bundle in self.bundles],
+            'binary_image': self.binary_image.to_json(),
+            'from_index': self.from_index.to_json(),
+            'user': self.user.username,
+            'latest_state': latest_state
+        }
+
+        if verbose:
+            rv['state_history'] = states
+
+        return rv
+
+    @classmethod
+    def from_json(cls, kwargs):
+        # Validate all required parameters are present
+        required_params = {'bundles', 'binary_image'}
+        optional_params = {'from_index', 'add_arches', 'cnr_token'}
+        missing_params = required_params - set(kwargs.keys()) - optional_params
+        if missing_params:
+            raise ValidationError('Missing required parameter(s): {}'
+                                  .format(', '.join(missing_params)))
+
+        # Don't allow the user to set arbitrary columns or relationships
+        invalid_params = set(kwargs.keys()) - required_params - optional_params
+        if invalid_params:
+            raise ValidationError(
+                'The following parameters are invalid: {}'.format(', '.join(invalid_params)))
+
+        request_kwargs = deepcopy(kwargs)
+
+        # Validate bundles are correctly provided
+        bundles = request_kwargs.get('bundles')
+        if not isinstance(bundles, list) or len(bundles) == 0:
+            raise ValidationError('"bundles" should be a non-empty array')
+
+        # Validate binary_image is correctly provided
+        if not isinstance(request_kwargs.get('binary_image'), str):
+            raise ValidationError('"binary_image" should be a string')
+
+        # Validate from_index and cnr_token if specified
+        for param in ('from_index', 'cnr_token'):
+            param_value = request_kwargs.get(param)
+            if param_value and not isinstance(param_value, str):
+                raise ValidationError(f'"{param}" must be a string')
+
+        # Validate add_arches are correctly provided
+        add_arches = request_kwargs.get('add_arches', [])
+        if not isinstance(add_arches, list):
+            raise ValidationError('"add_arches" should be an array')
+
+        # current_user.is_authenticated is only ever False when auth is disabled
+        if not current_user.is_authenticated:
+            raise ValidationError('Cannot set "user" when authentication is disabled')
+        request_kwargs['user_id'] = current_user.id
```

Comment:
Use `request_kwargs['user'] = current_user` instead if it works. I'd like to avoid interacting directly with the foreign key columns and let SQLAlchemy figure it out with the relationships we configured in the models.